### ISSUE TITLE
Startup/shutdown performance: unnecessary processing of activity events

### ISSUE THREAD

**Issue by HeikoKlare** [eclipse-platform/eclipse.platform.ui#1084]
## Issue
The activity management in `MutableActivityManager` and `ActivityPersistenceHelper` is inefficient when dealing with higher amounts of activities. Activity enablements are unnecessarily processed and persisted upon different events:
1. All activities enablements are persisted upon every activity enablement change in the preference store, even if only the enablement of an expression-controlled activity changes (which is not part of the persisted states):
https://github.com/eclipse-platform/eclipse.platform.ui/blob/f4989c93f6e599d977cabf7b039836a771a3b035/bundles/org.eclipse.ui.workbench/Eclipse%20UI/org/eclipse/ui/internal/ActivityPersistanceHelper.java#L60-L62
2. Enablement changes of expression-controlled activities are processed (and persisted as stated in 1.) even if the workbench is already closing. This has no effect as the activity support is disposed anyway during workbench close. Since the activity management is disposed _after_ the service locator, which produces further events due to the evaluation service being removed, it receives and processes these additional events unnecessarily.

## Reproduction

This unnecessary processing probably has no noticable effect in many products, as the types and amounts of activities are not high enough to produce a perceivable performance impact. In larger products this can, however, have severe impacts. In our product the effect is an overhead during startup and shutdown of the workbench of multiple seconds.

I have created a _very_ artificial demo project with a large amount of defined activities and a large amount of enablement changes during startup/shutdown to make the impact obvious:

[ManyActivitiesPerformanceDemo.zip](https://github.com/eclipse-platform/eclipse.platform.ui/files/12539744/ManyActivitiesPerformanceDemo.zip)

## Evaluation

I have executed that demo with the current state of implementation and compared it to improved implementations that address the mentioned issue (can be seen in PR #1085 and #1086).

On my machine, the results with the demo project across several runs were as follows:

Startup time:
Without unnecessary persistence: 350-500 ms
With unnecessary persistence: 1500-1700 ms

Shutdown time:
With early disposal: 350-450 ms
Without proper disposal: 1300-1450 ms

## How to proceed

I have started two PRs that address the mentioned issues (#1085, #1086). The changes of those PRs are the basis for the evaluation results mentioned above. In case you already see any side effects of addressing those issue that I am currently not aware of or, in general, reasons why the current behavior is required, I would be glad if you can let me know.

**Comment by vogella** [comment]
Thanks @HeikoKlare for working on this, improved startup performance is always welcomed. If you feel confident with your chanages, please merge them early in this release so that they get intensive usage during this development cycle.

**Comment by HeikoKlare** [comment]
Thanks @vogella! I will merge the PRs soon (probably this week). I will just find someone at least having a quick look at them first.

**Comment by fedejeanne** [comment]
I applied both of your PRs and these are my results

```
------------------------------------------------------------------
With PR-1085 and PR-1086
------------------------------------------------------------------
1st run
[With Activity Enablement Changes] Starting workbench took: 2242 ms.
[With Activity Enablement Changes] Closing workbench took: 3139 ms.
2nd run
[With Activity Enablement Changes] Starting workbench took: 2274 ms.
[With Activity Enablement Changes] Closing workbench took: 3103 ms.
3rd run
[With Activity Enablement Changes] Starting workbench took: 2316 ms.
[With Activity Enablement Changes] Closing workbench took: 3107 ms.
4th run
[With Activity Enablement Changes] Starting workbench took: 2430 ms.
[With Activity Enablement Changes] Closing workbench took: 3537 ms.
5th run
[With Activity Enablement Changes] Starting workbench took: 2509 ms.
[With Activity Enablement Changes] Closing workbench took: 3229 ms.

 

------------------------------------------------------------------
WIthout PRs
------------------------------------------------------------------
1st run
[With Activity Enablement Changes] Starting workbench took: 4321 ms.
[With Activity Enablement Changes] Closing workbench took: 6632 ms.
2nd run
[With Activity Enablement Changes] Starting workbench took: 4236 ms.
[With Activity Enablement Changes] Closing workbench took: 4537 ms.
3rd run
[With Activity Enablement Changes] Starting workbench took: 4191 ms.
[With Activity Enablement Changes] Closing workbench took: 4986 ms.
4th run
[With Activity Enablement Changes] Starting workbench took: 3829 ms.
[With Activity Enablement Changes] Closing workbench took: 5534 ms.
5th run
[With Activity Enablement Changes] Starting workbench took: 4299 ms.
[With Activity Enablement Changes] Closing workbench took: 4660 ms
```

Good performance improvement :-)

**Comment by HeikoKlare** [comment]
Thanks for that validation, @fedejeanne!
FWIW: the differences in the measurements provided in https://github.com/eclipse-platform/eclipse.platform.ui/issues/1084#issuecomment-1717760368 and the ones I provided in the initial posting https://github.com/eclipse-platform/eclipse.platform.ui/issues/1084#issue-1884297583 are not only due to different hardware but probably also because I reduced the set of plugins to those required for the test plugin, which reduces the overall execution times. When starting with a "complete" set of platform plugins, the measurements on my machine are comparable to those provided by @fedejeanne.
